composer/trainer/trainer.py

```diff
@@ -386,7 +384,7 @@ def create_from_hparams(cls, hparams: TrainerHparams) -> Trainer:
             callbacks=tuple(callbacks),
 
             # Checkpointing hparams
-            checkpoint_filepath=hparams.checkpoint_filepath,
+            checkpoint_loader=checkpoint_loader,
```

Comment:
I find this API a bit confusing .. checkpoint_loader is an hparams file that has a `checkpoint_filepath`, but the root trainer hparams also has `checkpoint_folder`?

For consistency, I would suggest not having a spearate nested hparams object, and keeping all the settings flat.

Reply:
Yes, this is because `CheckpointLoader` and `Checkpointer` are currently two different classes, where they should be one. So `checkpoint_folder` is for saving checkpoints, and `checkpoint_filepath` is for loading a checkpoint. 

To be clear, this is the previous API design, but I agree with your comments. 

@ravi-mosaicml thoughts if we should merge the hparams for saving and loading into a unified interface? 

This would broaden the scope of the PR (my goal was to get a blocker for BERT out of the way), but it probably is the correct thing to do.

Reply:
> For consistency, I would suggest not having a spearate nested hparams object, and keeping all the settings flat

Do we think this adds too many hparams to the base YAML? I would almost prefer something like:
```
load_checkpoint: 
   filepath: str
   load_weights_only: bool
   strict: bool
save_checkpoint:
   interval_unit: str
   folder: str
   interval: int
```

If we keep them flat, then the `folder` and `filepath` can become confusing, whereas one requires a folder (for saving multiple checkpoints) and the other requires the link to a single file (for loading a single checkpoint). This also doesn't clutter the main Trainer interface.

Additionally, `save_checkpoint.interval` should use the `Time` abstraction when it is ready.

Reply:
I would prefer to have a `checkpoint` section in the haparms where you can then have that above yaml structure.

Reply:
Yes, that is my thought. If both of you agree, I am happy to make the Trainer required changes. If we do that, can someone take care of addressing tests? I don't want to focus on this PR to the detriment of getting BERT ready in time.

Reply:
The save checkpoint won't require any additional tests. For load checkpoint with strict=False or load_weights_only=True, I think it would be trivial to add that in...I would be OK with a functional test (that just makes sure it runs, and doesn't necessarily assert that that warnings are raised or that only the model is loaded).

Reply:
I'm OK with this approach! Yes maybe @ravi-mosaicml if you aren't overloaded can help get this merged?

Reply:
@hanlint @ravi-mosaicml Awesome, I moved the checkpoint saving features into a `save_checkpoint` hyperparameter class. We now have two main level hparams: `load_checkpoint` and `save_checkpoint`, which have their respective hyperparameters.